### PR TITLE
Add max_comm_size to constrain the community size.

### PR DESCRIPTION
**Pull request by orenbenkiki**
First, thanks for publishing this package, and for supporting the "Surprise" goal function. I was working along these lines and just having it all be ready-made (and published) helped a lot.

I'm re-implementing the Metacell algorithm for analyzing single-cell RNA sequence data, and scaling it up to efficiently deal with very large data (many millions of cells) by converting it to a divide-and-conquer algorithm. I'm working on this as my PhD project in the Weizmann Institure for Science in Amos Tanay's lab.

This pull request implements a minor but important (for us) tweak to the basic clustering algorithm - restricting the maximal community size (sum of node sizes). Since metacells aren't "classical" clusters (in particular, they are defined as "the minimal sized community" that satisfies some conditions), this tweak makes it efficient to generate candidate metacells based on some KNN-graph, as part of the complete (complex) pipeline. I found that using this modified version saved CPU time and gave good results.

I tried to make this as seamless as possible with the existing code; two points where I had to tweak things were ensuring all vertex partitions allowed specifying node sizes, even if they did not directly use them, to allow passing them on to the algorithm itself; and change the parameters order for create_graph_from_py (since node_sizes became mandatory).

### CODE REVIEW

**Comment by orenbenkiki** [comment]
Just fixed a typo in the documentation - had "used used" in several places (the dangers of copy-and-paste), changed it to a single "used".

Happy new (Jewish) year!

**Comment by vtraag** [comment]
Hi, thanks! You too! Thanks for the PR btw, I am working on a review, I didn't get around to finishing it earlier on. I hope to get around to it this weekend somewhere...

**Review by vtraag — COMMENTED** [review]
Thanks for submitting this! It generally looks good, but I would change a few things, and I have some minor other comments.

Perhaps also a more general question: what do you think should happen if a node would be inside a community that is larger than `max_comm_size`? At the moment this is not caught, because it might be that moving it to a smaller community would actually decrease the quality. What do you think of this? I have some ideas for how to solve it, but I would be interested in hearing what you think.

File: include/python_partition_interface.h [hunk 1]
```diff
@@ -28,10 +28,9 @@ MutableVertexPartition* create_partition(Graph* graph, char* method, vector<size
 MutableVertexPartition* create_partition_from_py(PyObject* py_obj_graph, char* method, PyObject* py_initial_membership, PyObject* py_weights, PyObject* py_node_sizes, double resolution_parameter);
 
 Graph* create_graph_from_py(PyObject* py_obj_graph);
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply include/python_partition_interface.h]: You are right, the version w/o the node_sizes became redundant. I've removed it.

As for what happens if a node is inside a too-large partition....

TL;DR: I think the current pull-request stands on its own, and I'd be more than happy to discuss more ambitious follow-up work.

The (too long) answer is that I did think about it and my reasoning was as follows:

I viewed `max_comm_size` as an operational directive saying "do not add more nodes to _create_ too-large communities". This is a natural minor tweak to the algorithm, and if you apply this constraint to _all_ the phases, you are guaranteed to have not-too-large communities. Call this the "weak form" of the constraint. It seems useful on its own and can be used as a basic step for a more general approach.

I did not view `max_comm_size` as a more general directive of "ensure all communities are up to this size no matter what". I toyed with the idea, together with the "natural" accompanying `min_comm_size` to force the community sizes to be in some desired range. Call this the "strong form" of the constraint.

The "strong form" is basically what I need - metacells can't be either too small or too large. Right now I do this by using the "weak form" as a basic step.

I decided not to directly add "strong form" logic to the basic algorithm because this would no longer be a minor natural tweak to the algorithm. The current algorithm is a nice and simple bottom-up approach (ok, maybe not _that_ simple). But the optimal algorithm for establishing size constraints (especially given an initial partition) might no longer be a "simple" tweak of this approach.

I'm not certain how to _best_ split too-large communities, or combine too-small ones. I suspect this doesn't necessarily map to the bottom-up phased approach. The result might end up a "Leiden++" algorithm, which would be equivalent to the Leiden algorithm if the constraints would not be used.

Given that my need for tight control over the community size isn't a common one, I was not comfortable pushing such a major change into a general-purpose package - especially as the implementation would be rather "experimental".

If you are interested in pursuing this direction, I'd be happy to cooperate in doing so:

The "naive" method would be to force the move of a node to reduce the size of a too-large community, and likewise, force the move of nodes to increase the size of a too-small community.

The problem is that these two constraints conflict with each other, so there needs to be some balanced way to avoid endlessly splitting a too-large community into too-small ones, and merging these back into a too-large communities. This can be done by having some explicit global "penalty" function for violating the constraints, and being at least somewhat "greedy"
in minimizing it when making the forced moves.

From a bit of playing around with this, however, it seems that a purely greedy approach would not work well (which isn't very surprising). A better approach might be having an explicit "improve constraints" phase that used more global operations (e.g., "split too large a community into two").

BTW, my first approach was to use the unmodified Leiden algorithm, and recursively apply it to only "too-large" communities to split them into smaller parts. Then apply the unmodified algorithm to a graph where each node is a too-small community, to merge them into larger communities, going back and forth as long as I improved the overall result.

This worked, mostly, but being able to avoid the "split too large communities" step by tweaking the basic algorithm to not generate them in the 1st place turned out to be a significant improvement (both in CPU time and quality of the result). So, what I do now is use `max_comm_size` to restrict the community size. "All" I need to do next is to do the "merge too-small communities" phase on the result. This is much simpler and seems to work well.

It seems reasonable that a more general algorithm that deals with both sides of the constraints would do even better. I did make an attempt along these lines but it didn't work too well (probably due to being too "greedy"). This is still something I might want to pursue.

In the meanwhile, I'd like to have the simple `max_comm_size` tweak in the `ledenalg` package, so I can publish my `metacells` package that depends on it, so users can start playing with it.

Sorry for the rambling answer...

File: src/python_optimiser_interface.cpp [hunk 2]
```diff
@@ -40,16 +40,18 @@ extern "C"
...
 
-    static char* kwlist[] = {"optimiser", "partition", "is_membership_fixed", NULL};
+    static char* kwlist[] = {"optimiser", "partition", "is_membership_fixed", "max_comm_size", NULL};
 
     #ifdef DEBUG
       cerr << "Parsing arguments..." << endl;
     #endif
 
-    if (!PyArg_ParseTupleAndKeywords(args, keywds, "OO|O", kwlist,
+    if (!PyArg_ParseTupleAndKeywords(args, keywds, "OO|OI", kwlist,
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply src/python_optimiser_interface.cpp]: You are right; however, I just removed this parameter from all these calls (due to your comment below) so this became a moot issue.

I did add an explicit test in the Python side when setting `max_comm_size` to ensure negative values are reported as an error. I was hoping the `I` keyword parser would do this automatically, it turns out it does not.

File: src/Optimiser.cpp [hunk 3]
```diff
@@ -605,6 +642,11 @@ double Optimiser::move_nodes(vector<MutableVertexPartition*> partitions, vector<
       {
         graph = graphs[layer];
         partition = partitions[layer];
+        // Do not create too-large communities.
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply src/Optimiser.cpp]: Good point. I introduced `v_size` before the `comm` loop and moved the `max_comm_size` test to the top body. Less code, more efficient. Thanks!

File: src/Optimiser.py [hunk 4]
```diff
@@ -229,7 +245,7 @@ def set_rng_seed(self, value):
     """
     _c_leiden._Optimiser_set_rng_seed(self._optimiser, value)
 
-  def optimise_partition(self, partition, n_iterations=2, is_membership_fixed=None):
+  def optimise_partition(self, partition, n_iterations=2, is_membership_fixed=None, max_comm_size=None):
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply src/Optimiser.py]: No, you are right. I removed the parameter from all these methods. It is easy enough to just set `max_comm_size = ...` before the call.

File: src/Optimiser.py [hunk 5]
```diff
@@ -247,6 +263,10 @@ def optimise_partition(self, partition, n_iterations=2, is_membership_fixed=None
       length of this list must be equal to the number of nodes. By default
       (None) all nodes can change community during the optimization.
 
+    max_comm_size : int or None
+      Maximal total size of nodes in a community. If zero, then communities can
+      be of any size. If ``None`` (the default), uses uses :attr:`max_comm_size`.
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply src/Optimiser.py]: I thought I caught all of these... at any rate, since I removed the parameter, this became a moot issue.

File: src/functions.py [hunk 6]
```diff
@@ -23,7 +23,7 @@ def _get_py_capsule(graph):
 from .VertexPartition import *
 from .Optimiser import *
 
-def find_partition(graph, partition_type, initial_membership=None, weights=None, n_iterations=2, seed=None, **kwargs):
+def find_partition(graph, partition_type, initial_membership=None, weights=None, n_iterations=2, seed=None, max_comm_size=0, **kwargs):
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply src/functions.py]: No, I think you were correct above. Using `max_comm_size` is rare enough that I shouldn't have inflicted it on the API of all these functions. I took it out here as well.

File: tests/test_Optimiser.py [hunk 7]
```diff
@@ -21,6 +21,14 @@ def test_move_nodes(self):
         partition.sizes(), [100],
         msg="CPMVertexPartition(resolution_parameter=0.5) of complete graph after move nodes incorrect.");
 
+  def test_move_nodes_with_max_comm_size(self):
+    G = ig.Graph.Full(100);
+    partition = leidenalg.CPMVertexPartition(G, resolution_parameter=0.5);
+    self.optimiser.move_nodes(partition, consider_comms=leidenalg.ALL_NEIGH_COMMS, max_comm_size=17);
+    self.assertListEqual(
+        partition.sizes(), [17, 17, 17, 17, 17, 15],
+        msg="CPMVertexPartition(resolution_parameter=0.5) of complete graph after move nodes incorrect.");
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply tests/test_Optimiser.py]: Done.

File: tests/test_Optimiser.py [hunk 8]
```diff
@@ -45,6 +53,14 @@ def test_merge_nodes(self):
         G.ecount(),
         msg="total_weight_in_all_comms not equal to ecount of graph.");
 
+  def test_merge_nodes_with_max_comm_size(self):
+    G = ig.Graph.Full(100);
+    partition = leidenalg.CPMVertexPartition(G, resolution_parameter=0.5);
+    self.optimiser.merge_nodes(partition, consider_comms=leidenalg.ALL_NEIGH_COMMS, max_comm_size=17);
+    self.assertListEqual(
+        partition.sizes(), [17, 17, 17, 17, 17, 15],
+        msg="CPMVertexPartition(resolution_parameter=0.5) of complete graph after merge nodes incorrect.");
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply tests/test_Optimiser.py]: Done.

File: tests/test_Optimiser.py [hunk 9]
```diff
@@ -66,6 +82,15 @@ def test_optimiser(self):
...
         msg="After optimising partition failed to find different components with CPMVertexPartition(resolution_parameter=0)");
 
+  def test_optimiser_with_max_comm_size(self):
+    G = ig.Graph.Full(100);
+    partition = leidenalg.CPMVertexPartition(G, resolution_parameter=0);
+    self.optimiser.consider_comms=leidenalg.ALL_NEIGH_COMMS;
+    self.optimiser.optimise_partition(partition, max_comm_size=10);
+    self.assertListEqual(
+        partition.sizes(), 10*[10],
+        msg="After optimising partition failed to find different components with CPMVertexPartition(resolution_parameter=0)");
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply tests/test_Optimiser.py]: Done.

**Comment by orenbenkiki** [comment]
Fixed all issues (I think) as far as this "weak form" pull request goes.
As for the "strong form", I suggest that if we want to pursue it, it should be done not here but in a dedicated "issue".

Thanks for the review!

**Comment by vtraag** [comment]
Looks good, thanks for the quick response!

I'll think about your comment on the "strong form", perhaps the solution I have in mind also does not work well. The idea I had was quite simple in fact: we do not set the current community as the maximum one, and we set the initial maximum improvement to -infinity, while adding the current community to the set of  communities to consider. That way, if the current community is too large it will not be considered. There will always be a possibility to move it to an empty community (assuming that option is turned on). What do you think?

We also have to deal with too small clusters in our work (clustering scientific publications). In that case, we post-process to join too small communities with larger communities, such that is minimises the loss of forcing the solution to exclude too small communities. If you are interested, I can talk you through this. It is more difficult to enforce this during the running of the algorithm however, since initially all communities are too small.

**Comment by orenbenkiki** [comment]
Thanks for the quick reply.

> I'll think about your comment on the "strong form", perhaps the solution I have in mind also does not work well. The idea I had was quite simple in fact: we do not set the current community as the maximum one, and we set the initial maximum improvement to -infinity, while adding the current community to the set of communities to consider. That way, if the current community is too large it will not be considered. There will always be a possibility to move it to an empty community (assuming that option is turned on). What do you think?

So, if I get this right, this effectively will split the 1st several nodes from the too-large community, forcing them to either join other existing or create new communities. An elegant solution and easy to implement.

There's a nasty edge case when all nodes belong to a single too-large community and we are not allowed to create new communities, though.

That aside, I'm a _bit_ worried in that the nodes chosen to be split from the too-large community are essentially chosen at random, and this may impact the optimality / efficiency of the algorithm.

A more serious potential concern is if we have "very" too-large communities (e.g., several times the maximal size). It might be more efficient to run the overall algorithm (or a simplified version of it) on the too-large community to split it into several smaller ones? If that is an overkill, perhaps even a simple "split to random subsets of roughly equal size" step might end up being more efficient?

> We also have to deal with too small clusters in our work (clustering scientific publications). In that case, we post-process to join too small communities with larger communities, such that is minimises the loss of forcing the solution to exclude too small communities. If you are interested, I can talk you through this. It is more difficult to enforce this during the running of the algorithm however, since initially all communities are too small.

What I do right now is to convert each too-small community into a node, and run the algorithm on the resulting graph (containing only the too-small community nodes) to cluster them into larger ones. This seems to be "in the spirit" of the overall algorithm. Typically it solves the problem; sometimes I am left with "stubborn" too-small communities which really resist being clustered with anything else, which I basically consider to contain "outlier" nodes.

I would be interested in a better approach - as you say, I discovered it is difficult to enforce this during the algorithm (creating new communities becomes an issue).

File: src/functions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply src/functions.py]: Correction: It _does_ make sense to pass `max_com_size` to these functions since they create an internal `Optimiser` and there's no easy way to tweak it short of an explicit parameter. That said, I added an explicit check for negative values to the code.

[the remark this answers is not in the document]
Comment (vtraag) [reply src/functions.py]: The same hence goes for other settings (e.g. `consider_empty_communtiy`), but as said, I think we can allow an exception for this case.

File: src/python_optimiser_interface.cpp [hunk 10]
```diff
@@ -871,6 +871,70 @@ extern "C"
     return PyBool_FromLong(optimiser->refine_partition);
   }
 
+  PyObject* _Optimiser_set_max_comm_size(PyObject *self, PyObject *args, PyObject *keywds)
+  {
+    PyObject* py_optimiser = NULL;
+    unsigned int max_comm_size = 0;
```
Comment (vtraag) [inline src/python_optimiser_interface.cpp:877]: The package actually uses `size_t` to represent unsigned integers. This ensures it is correctly compiled for 64-bit support (`int` is 32 bit), this is also how `max_comm_size` is defined in the actual C optimiser object.

File: src/python_optimiser_interface.cpp [hunk 11]
```diff
@@ -871,6 +871,70 @@ extern "C"
...
+    static char* kwlist[] = {"optimiser", "max_comm_size", NULL};
+
+    #ifdef DEBUG
+      cerr << "Parsing arguments..." << endl;
+    #endif
+
+    if (!PyArg_ParseTupleAndKeywords(args, keywds, "OI", kwlist,
```
Comment (vtraag) [inline src/python_optimiser_interface.cpp:884]: For reading a `size_t` object, you would need the specifier `n`.

File: src/python_optimiser_interface.cpp [hunk 12]
```diff
@@ -871,6 +871,70 @@ extern "C"
...
+    #ifdef DEBUG
+      cerr << "Capsule optimiser at address " << py_optimiser << endl;
+    #endif
+    Optimiser* optimiser = decapsule_Optimiser(py_optimiser);
+    #ifdef DEBUG
+      cerr << "Using optimiser at address " << optimiser << endl;
+    #endif
+
+    #ifdef IS_PY3K
+    return PyLong_FromLong(optimiser->max_comm_size);
```
Comment (vtraag) [inline src/python_optimiser_interface.cpp:932]: Here, to translate from a `size_t` you need `PyLong_FromSize_t`, this works for both Python 2 and Python 3.

**Review by vtraag — COMMENTED** [review]
Thanks again for submitting the fixes.

I think it is good to also try to enforce the `max_comm_size` on the current community, just for consistency. I agree that enforcing this constraints works better when starting from a singleton partition, growing the communities up until the `max_comm_size` limit, than when having to shave off single nodes one by one from a too large community. Thanks for including this.

There is one small minor issue dealing with the correct type of the `max_comm_size` issue. After that, I'd be happy to merge this in! After this is merged in, I will also make a new release.

File: src/python_optimiser_interface.cpp (hunk 11 above)
Comment (orenbenkiki) [reply src/python_optimiser_interface.cpp]: Done.

File: src/python_optimiser_interface.cpp (hunk 12 above)
Comment (orenbenkiki) [reply src/python_optimiser_interface.cpp:932]: Just to make sure - this means I should use `PyLong_FromSize_t` w/o an `#ifdef`?

Comment (vtraag) [reply src/python_optimiser_interface.cpp]: Good point that elsewhere it could also use `PyLong_FromLong`, not sure why that is still in there. Simply use unconditionally `PyLong_FromSize_t` should be fine, thanks!

**Comment by vtraag** [comment]
Thanks! I will merge after the CI has finished.

**Comment by orenbenkiki** [comment]
Thanks!
